core/src/main/java/jenkins/security/auth/LoginThrottler.java

```diff
@@ -0,0 +1,206 @@
+import org.acegisecurity.userdetails.UserDetails;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+@Extension
+public class LoginThrottler extends SecurityListener {
+    private static final Logger LOGGER = Logger.getLogger(LoginThrottler.class.getName());
+
+    private static final Integer RATE_LIMITER_WINDOW_DURATION = Integer.getInteger(LoginThrottler.class.getName() + ".duration", 30);
+    private static final Integer RATE_LIMITER_MAX_ATTEMPTS = Integer.getInteger(LoginThrottler.class.getName() + ".attempts", 3);
+
+    private ReadWriteLock failuresLock = new ReentrantReadWriteLock();
+    private Map<String, UserLoginFailuresInfo> failures = new HashMap<>();
+    private boolean active = true;
+
+    private static class UserLoginFailuresInfo {
+        private SortedSet<Date> failures;
+
+        private String name;
+
+        private UserLoginFailuresInfo(String name) {
+            failures = new TreeSet<>();
+            this.name = name;
+        }
+
+        public void recordFailure() {
+            LOGGER.log(Level.FINE, () -> "Recording login failure for " + name);
+
+            long nowTime = TimeProvider.get().getCurrentTimeMillis();
+            Date now = new Date(nowTime);
+            failures.add(now);
+
+            // remove login failures from earlier than the relevant rate limiting window
+            Date ago = new Date(nowTime);
+            ago.setTime(ago.getTime() - RATE_LIMITER_WINDOW_DURATION * 1000);
+            failures = failures.tailSet(ago);
+
+            LOGGER.log(Level.FINE, () -> "There are now " + failures.size() + " recorded login failures for " + name);
+        }
+
+        public boolean isSevereFailureRate() {
+            long nowTime = TimeProvider.get().getCurrentTimeMillis();
+            Date ago = new Date(nowTime);
+            ago.setTime(ago.getTime() - RATE_LIMITER_WINDOW_DURATION * 1000);
+            int failureCountInTimeSpan = failures.tailSet(ago).size();
+
+            LOGGER.log(Level.FINE, () -> "Determining failure rate for " + name + " at " + failureCountInTimeSpan + " of " + RATE_LIMITER_MAX_ATTEMPTS + " in " + RATE_LIMITER_WINDOW_DURATION + " seconds since: " + ago.toString());
+
+            if (failureCountInTimeSpan >= RATE_LIMITER_MAX_ATTEMPTS) {
+                return true;
+            }
+            return false;
+        }
+    }
+
+    public static LoginThrottler get() {
+        return Jenkins.get().getExtensionList(SecurityListener.class).get(LoginThrottler.class);
+    }
+
+    public void setActive(boolean active) {
```

Comment:
Mark this for tests only.

There are also other methods here that are currently only used for tests. If unchanged they need to be marked for tests only.